// impor/out/index.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const path_1 = require("path");
function impor(id, rootDir) {
    if (/^\./.test(id)) {
        id = path_1.join(rootDir, id);
    }
    if (ImportJob.importQueue.indexOf(id) === -1) {
        ImportJob.importQueue.push(id);
    }
    let mod;
    return new Proxy((() => { }), {
        apply: (_, thisArg, argumentsList) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.apply(mod, thisArg, argumentsList);
        },
        construct: (_, argumentsList) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.construct(mod, argumentsList);
        },
        defineProperty: (_, key, attributes) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.defineProperty(mod, key, attributes);
        },
        deleteProperty: (_, key) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.deleteProperty(mod, key);
        },
        get: (_, key, reciver) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.get(mod, key, reciver);
        },
        getPrototypeOf: (_) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.getPrototypeOf(mod);
        },
        has: (_, key) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.has(mod, key);
        },
        isExtensible: (_) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.isExtensible(mod);
        },
        ownKeys: (_) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return mod ? Reflect.ownKeys(mod) : [];
        },
        preventExtensions: (_) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.preventExtensions(mod);
        },
        set: (_, key, value, reciver) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.set(mod, key, value, reciver);
        },
        setPrototypeOf: (_, value) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.setPrototypeOf(mod, value);
        }
    });
}
class ImportJob {
    static importJob() {
        return __awaiter(this, void 0, void 0, function* () {
            if (ImportJob.importJobBusy) {
                return;
            }
            if (ImportJob.importQueue.length === 0) {
                return;
            }
            const id = ImportJob.importQueue.shift();
            if (!id) {
                return;
            }
            ImportJob.importJobBusy = true;
            yield Promise.resolve().then(() => require(id));
            ImportJob.importJobBusy = false;
        });
    }
    static removeJob(id) {
        const index = ImportJob.importQueue.indexOf(id);
        if (index === -1) {
            return;
        }
        ImportJob.importQueue.splice(index, 1);
    }
    static start() {
        if (ImportJob.isStarted) {
            return;
        }
        ImportJob.isStarted = true;
        setInterval(ImportJob.importJob, 1000);
    }
}
ImportJob.isStarted = false;
ImportJob.importQueue = [];
ImportJob.importJobBusy = false;
ImportJob.start();
module.exports = (rootDir) => {
    return (id) => {
        return impor(id, rootDir);
    };
};
// SIG // Begin signature block
// SIG // MIInygYJKoZIhvcNAQcCoIInuzCCJ7cCAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // sygLtaT7wF2/NPBw/4MdTDHKba4WjOQvLRL9afPvJZyg
// SIG // gg2WMIIGFDCCA/ygAwIBAgITMwAAAkNVsbk2MwXnzgAA
// SIG // AAACQzANBgkqhkiG9w0BAQsFADB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExMB4XDTIxMDgxMjE3MzgzNFoX
// SIG // DTIyMDkxNTE3MzgzNFowgYgxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xMjAwBgNVBAMTKU1pY3Jvc29mdCAzcmQgUGFydHkg
// SIG // QXBwbGljYXRpb24gQ29tcG9uZW50MIIBIjANBgkqhkiG
// SIG // 9w0BAQEFAAOCAQ8AMIIBCgKCAQEAu2BwjeXSQdxF6x8F
// SIG // 1GWoEvK4ztmQe1eLQs9GDzilmb9Sgt3qFCifIS5EKiDg
// SIG // zHtqKbAyLcsNDSz/TceaKy+HyqMalJhIVmWwxdZ0Fed3
// SIG // M69bBiBwK+gtWBitqhZIjzZSs+Oovp1hrsxJLygh7XML
// SIG // 5DiCdgpDC7OTf6npJ+rF6QrTSaYMMSBKHfnbj+xkcAfh
// SIG // yqhTxU6z1UOq7ERa4Bmuscxnegyudy4MPAeTbde6QD50
// SIG // QiLQkD0T7iYhy2JARvc1TOhnG92zn4x085S948D3MbMk
// SIG // aOQNHLvcJGGTsjv4d1NUhhtvWYYlkykxTXMQLVtz0XZX
// SIG // oN4iJ5S+lLJeUHDwCQIDAQABo4IBfjCCAXowHwYDVR0l
// SIG // BBgwFgYKKwYBBAGCN0wRAQYIKwYBBQUHAwMwHQYDVR0O
// SIG // BBYEFOqaLrRdpqfTP5zrgAiy+2S3xx1FMFAGA1UdEQRJ
// SIG // MEekRTBDMSkwJwYDVQQLEyBNaWNyb3NvZnQgT3BlcmF0
// SIG // aW9ucyBQdWVydG8gUmljbzEWMBQGA1UEBRMNMjMxNTIy
// SIG // KzQ2NjQxMjAfBgNVHSMEGDAWgBRIbmTlUAXTgqoXNzci
// SIG // tW2oynUClTBUBgNVHR8ETTBLMEmgR6BFhkNodHRwOi8v
// SIG // d3d3Lm1pY3Jvc29mdC5jb20vcGtpb3BzL2NybC9NaWND
// SIG // b2RTaWdQQ0EyMDExXzIwMTEtMDctMDguY3JsMGEGCCsG
// SIG // AQUFBwEBBFUwUzBRBggrBgEFBQcwAoZFaHR0cDovL3d3
// SIG // dy5taWNyb3NvZnQuY29tL3BraW9wcy9jZXJ0cy9NaWND
// SIG // b2RTaWdQQ0EyMDExXzIwMTEtMDctMDguY3J0MAwGA1Ud
// SIG // EwEB/wQCMAAwDQYJKoZIhvcNAQELBQADggIBAEfTGT+m
// SIG // Q+BEs2cxilP1tH5emLonNAFlif4s/5TVgOt6wAODVMpW
// SIG // wyGdSRZyMK4Z1OWsR5h2y/qEsFSyk1sPqLw4+ZkaLnEy
// SIG // +taAEbT/0EDHaOuJa2XP6vMXhPQPCxJpAz7IzWyCzW1F
// SIG // EUy5fN5TlWp2MKLdrSkbV7euXoa+uftZDk3fFlrKIewQ
// SIG // yriRPgrRuqJcKd5U5oq4FLDUzAldvwiIdhO7jdwVI3ct
// SIG // jSkh4VuS+mytysrL3Ok1u5NN4Mt0OlYcGpNbNibLljwx
// SIG // RPNwOaolrrGNcNLxXC7bQWnOQlJtcbmh82NJn31f4uH1
// SIG // TK6EmJej82BdbPsxLDZkaCMouLW37HPm0R1refE2u4H4
// SIG // lBnBiIza/NY8yPYTDWwt1N1sQ7o8uh2PmoW2LOOhLWvO
// SIG // Hm7YtbuIe/RDAmQljocDtlxyVMMDY19fiVRJban/tLgm
// SIG // qUnoWJ/oujew++p5nr64QQPtZhm5khRkJqht+FtPVMPb
// SIG // qjlGvcTL73xMs9v2UYSQTWgoZJMJVpcY+wikePYY1cHq
// SIG // 7T5Io7p65f5T4a8knxPOCrrNYaCg0hash4QgnB7f7c8W
// SIG // rNyGhA2FJtHYNmDEnjQBejkauoGrm0xl8DmNRbCK9dRN
// SIG // cPptaJc3sLsLRqu3Q36Wz7vzWYZWCPod87Q0mZByox01
// SIG // Jrluh/j0EpvL9vPKMIIHejCCBWKgAwIBAgIKYQ6Q0gAA
// SIG // AAAAAzANBgkqhkiG9w0BAQsFADCBiDELMAkGA1UEBhMC
// SIG // VVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcT
// SIG // B1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jw
// SIG // b3JhdGlvbjEyMDAGA1UEAxMpTWljcm9zb2Z0IFJvb3Qg
// SIG // Q2VydGlmaWNhdGUgQXV0aG9yaXR5IDIwMTEwHhcNMTEw
// SIG // NzA4MjA1OTA5WhcNMjYwNzA4MjEwOTA5WjB+MQswCQYD
// SIG // VQQGEwJVUzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4G
// SIG // A1UEBxMHUmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0
// SIG // IENvcnBvcmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQg
// SIG // Q29kZSBTaWduaW5nIFBDQSAyMDExMIICIjANBgkqhkiG
// SIG // 9w0BAQEFAAOCAg8AMIICCgKCAgEAq/D6chAcLq3YbqqC
// SIG // EE00uvK2WCGfQhsqa+laUKq4BjgaBEm6f8MMHt03a8YS
// SIG // 2AvwOMKZBrDIOdUBFDFC04kNeWSHfpRgJGyvnkmc6Whe
// SIG // 0t+bU7IKLMOv2akrrnoJr9eWWcpgGgXpZnboMlImEi/n
// SIG // qwhQz7NEt13YxC4Ddato88tt8zpcoRb0RrrgOGSsbmQ1
// SIG // eKagYw8t00CT+OPeBw3VXHmlSSnnDb6gE3e+lD3v++Mr
// SIG // WhAfTVYoonpy4BI6t0le2O3tQ5GD2Xuye4Yb2T6xjF3o
// SIG // iU+EGvKhL1nkkDstrjNYxbc+/jLTswM9sbKvkjh+0p2A
// SIG // LPVOVpEhNSXDOW5kf1O6nA+tGSOEy/S6A4aN91/w0FK/
// SIG // jJSHvMAhdCVfGCi2zCcoOCWYOUo2z3yxkq4cI6epZuxh
// SIG // H2rhKEmdX4jiJV3TIUs+UsS1Vz8kA/DRelsv1SPjcF0P
// SIG // UUZ3s/gA4bysAoJf28AVs70b1FVL5zmhD+kjSbwYuER8
// SIG // ReTBw3J64HLnJN+/RpnF78IcV9uDjexNSTCnq47f7Fuf
// SIG // r/zdsGbiwZeBe+3W7UvnSSmnEyimp31ngOaKYnhfsi+E
// SIG // 11ecXL93KCjx7W3DKI8sj0A3T8HhhUSJxAlMxdSlQy90
// SIG // lfdu+HggWCwTXWCVmj5PM4TasIgX3p5O9JawvEagbJjS
// SIG // 4NaIjAsCAwEAAaOCAe0wggHpMBAGCSsGAQQBgjcVAQQD
// SIG // AgEAMB0GA1UdDgQWBBRIbmTlUAXTgqoXNzcitW2oynUC
// SIG // lTAZBgkrBgEEAYI3FAIEDB4KAFMAdQBiAEMAQTALBgNV
// SIG // HQ8EBAMCAYYwDwYDVR0TAQH/BAUwAwEB/zAfBgNVHSME
// SIG // GDAWgBRyLToCMZBDuRQFTuHqp8cx0SOJNDBaBgNVHR8E
// SIG // UzBRME+gTaBLhklodHRwOi8vY3JsLm1pY3Jvc29mdC5j
// SIG // b20vcGtpL2NybC9wcm9kdWN0cy9NaWNSb29DZXJBdXQy
// SIG // MDExXzIwMTFfMDNfMjIuY3JsMF4GCCsGAQUFBwEBBFIw
// SIG // UDBOBggrBgEFBQcwAoZCaHR0cDovL3d3dy5taWNyb3Nv
// SIG // ZnQuY29tL3BraS9jZXJ0cy9NaWNSb29DZXJBdXQyMDEx
// SIG // XzIwMTFfMDNfMjIuY3J0MIGfBgNVHSAEgZcwgZQwgZEG
// SIG // CSsGAQQBgjcuAzCBgzA/BggrBgEFBQcCARYzaHR0cDov
// SIG // L3d3dy5taWNyb3NvZnQuY29tL3BraW9wcy9kb2NzL3By
// SIG // aW1hcnljcHMuaHRtMEAGCCsGAQUFBwICMDQeMiAdAEwA
// SIG // ZQBnAGEAbABfAHAAbwBsAGkAYwB5AF8AcwB0AGEAdABl
// SIG // AG0AZQBuAHQALiAdMA0GCSqGSIb3DQEBCwUAA4ICAQBn
// SIG // 8oalmOBUeRou09h0ZyKbC5YR4WOSmUKWfdJ5DJDBZV8u
// SIG // LD74w3LRbYP+vj/oCso7v0epo/Np22O/IjWll11lhJB9
// SIG // i0ZQVdgMknzSGksc8zxCi1LQsP1r4z4HLimb5j0bpdS1
// SIG // HXeUOeLpZMlEPXh6I/MTfaaQdION9MsmAkYqwooQu6Sp
// SIG // BQyb7Wj6aC6VoCo/KmtYSWMfCWluWpiW5IP0wI/zRive
// SIG // /DvQvTXvbiWu5a8n7dDd8w6vmSiXmE0OPQvyCInWH8My
// SIG // GOLwxS3OW560STkKxgrCxq2u5bLZ2xWIUUVYODJxJxp/
// SIG // sfQn+N4sOiBpmLJZiWhub6e3dMNABQamASooPoI/E01m
// SIG // C8CzTfXhj38cbxV9Rad25UAqZaPDXVJihsMdYzaXht/a
// SIG // 8/jyFqGaJ+HNpZfQ7l1jQeNbB5yHPgZ3BtEGsXUfFL5h
// SIG // YbXw3MYbBL7fQccOKO7eZS/sl/ahXJbYANahRr1Z85el
// SIG // CUtIEJmAH9AAKcWxm6U/RXceNcbSoqKfenoi+kiVH6v7
// SIG // RyOA9Z74v2u3S5fi63V4GuzqN5l5GEv/1rMjaHXmr/r8
// SIG // i+sLgOppO6/8MO0ETI7f33VtY5E90Z1WTk+/gFcioXgR
// SIG // MiF670EKsT/7qMykXcGhiJtXcVZOSEXAQsmbdlsKgEhr
// SIG // /Xmfwb1tbWrJUnMTDXpQzTGCGYwwghmIAgEBMIGVMH4x
// SIG // CzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9u
// SIG // MRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNy
// SIG // b3NvZnQgQ29ycG9yYXRpb24xKDAmBgNVBAMTH01pY3Jv
// SIG // c29mdCBDb2RlIFNpZ25pbmcgUENBIDIwMTECEzMAAAJD
// SIG // VbG5NjMF584AAAAAAkMwDQYJYIZIAWUDBAIBBQCgga4w
// SIG // GQYJKoZIhvcNAQkDMQwGCisGAQQBgjcCAQQwHAYKKwYB
// SIG // BAGCNwIBCzEOMAwGCisGAQQBgjcCARUwLwYJKoZIhvcN
// SIG // AQkEMSIEIEoNofzyFv/FYdYpXA3vZWXbWsOj0EBBI1/S
// SIG // 5olBN14BMEIGCisGAQQBgjcCAQwxNDAyoBSAEgBNAGkA
// SIG // YwByAG8AcwBvAGYAdKEagBhodHRwOi8vd3d3Lm1pY3Jv
// SIG // c29mdC5jb20wDQYJKoZIhvcNAQEBBQAEggEAiL8m/Cx4
// SIG // /SkdQZlFoLIrr0MB+NKq2l7mDD7ZpAFgyZNYg5mtWnFG
// SIG // yln+Hkd+XXFXl6DYOn58OmlL0dhWZ50Yca3Xv3BqjUXD
// SIG // 9B9OUe0cUgcqqWnccxDw9YWx8y24sHQOrLEWeWX79Flb
// SIG // CBQYUgV6UoKDibcZglBeGflrUssSMa8yZuow27V/XSy1
// SIG // nyroCDd2X7OAWhbKeLQh2JJ7WbSFUTIiuyI4h2jGHysA
// SIG // D1sXJKfuHqLBhAjqKae6OU8apjgglkhYZUtXBtkuz+lD
// SIG // hrTDdLcwhficxB+/0LXqhw2WDislyokl7xbt6vn0YpGN
// SIG // gAzBcVc7tVC02SIiXhZuAhk+NqGCFxYwghcSBgorBgEE
// SIG // AYI3AwMBMYIXAjCCFv4GCSqGSIb3DQEHAqCCFu8wghbr
// SIG // AgEDMQ8wDQYJYIZIAWUDBAIBBQAwggFZBgsqhkiG9w0B
// SIG // CRABBKCCAUgEggFEMIIBQAIBAQYKKwYBBAGEWQoDATAx
// SIG // MA0GCWCGSAFlAwQCAQUABCAQ3R+ze0kJU5otRu/nN8vf
// SIG // KlJ/emm7JOHj6+yMNllmoQIGYmxFRHTiGBMyMDIyMDUw
// SIG // MzE5MDEwOC42MjlaMASAAgH0oIHYpIHVMIHSMQswCQYD
// SIG // VQQGEwJVUzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4G
// SIG // A1UEBxMHUmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0
// SIG // IENvcnBvcmF0aW9uMS0wKwYDVQQLEyRNaWNyb3NvZnQg
// SIG // SXJlbGFuZCBPcGVyYXRpb25zIExpbWl0ZWQxJjAkBgNV
// SIG // BAsTHVRoYWxlcyBUU1MgRVNOOjA4NDItNEJFNi1DMjlB
// SIG // MSUwIwYDVQQDExxNaWNyb3NvZnQgVGltZS1TdGFtcCBT
// SIG // ZXJ2aWNloIIRZTCCBxQwggT8oAMCAQICEzMAAAGHQhZm
// SIG // BFzz9I0AAQAAAYcwDQYJKoZIhvcNAQELBQAwfDELMAkG
// SIG // A1UEBhMCVVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAO
// SIG // BgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29m
// SIG // dCBDb3Jwb3JhdGlvbjEmMCQGA1UEAxMdTWljcm9zb2Z0
// SIG // IFRpbWUtU3RhbXAgUENBIDIwMTAwHhcNMjExMDI4MTky
// SIG // NzM5WhcNMjMwMTI2MTkyNzM5WjCB0jELMAkGA1UEBhMC
// SIG // VVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcT
// SIG // B1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jw
// SIG // b3JhdGlvbjEtMCsGA1UECxMkTWljcm9zb2Z0IElyZWxh
// SIG // bmQgT3BlcmF0aW9ucyBMaW1pdGVkMSYwJAYDVQQLEx1U
// SIG // aGFsZXMgVFNTIEVTTjowODQyLTRCRTYtQzI5QTElMCMG
// SIG // A1UEAxMcTWljcm9zb2Z0IFRpbWUtU3RhbXAgU2Vydmlj
// SIG // ZTCCAiIwDQYJKoZIhvcNAQEBBQADggIPADCCAgoCggIB
// SIG // AL5peBljPQOj0REIh4GQADyup/k4epvi2MCmC2kBZWV2
// SIG // 06q6HYbJ4ux1SEZKgRbhvKINSma8tNsRJpm3afPTuNMQ
// SIG // iEH1vBSlu6Dv5WGLVAEFtpKdLjLmBCdV5Genh43DLqhC
// SIG // cAuO1GpGcdwHpGpD/hnxrgwzvZGRkEXvCKYoieEJJ+pm
// SIG // w4W0aYHFDnbA85jc+pFxu4HyH4HLOZM7iiD+pzeLbIz4
// SIG // chefFVNGXyND0gvp+Tl9IbFPIXlnt6VIg6+YKj9emupC
// SIG // IxSR5IDT2IDIrxk0CnIzcTXF558pk/yVbjbO2b5NT+Sl
// SIG // n/BtVxvzLWWef47qrINEKFJkGY1YdF0ZLu2KTKZLVu13
// SIG // xQN8vmCThtaVc4swdz8yV1AvDGvAFMsLVLlC1o2E+cu/
// SIG // xMMB1pp9UU9D4WKfZJqX0uhyGHxAW6GFL1C1th5xjX8e
// SIG // uX+FJHO5IDUkG3gB+sozuXYOjFMD386P6wXTWmXycWe0
// SIG // bHEwck7ZcW5X1E+OsjxwlwAol/wYWhxGOHnRYKf3CG+1
// SIG // xxSW7uJjQwD80ZQJCI9UhrjvRuLG5IzPV6TTFoJY9nQv
// SIG // UCzf9IHgRV6DSPBfh3x8OCaPqeokj/IZ+N5pJFikhjok
// SIG // +ISyc6HBKxZ7SuYvT3m10qtgbzIM1XPgQbwjd+Q3w9Jq
// SIG // JESsWptrUDNOtEjkct0u7nHrAgMBAAGjggE2MIIBMjAd
// SIG // BgNVHQ4EFgQUIlHAFrpBLWGBft1hc5mQPw2ljSAwHwYD
// SIG // VR0jBBgwFoAUn6cVXQBeYl2D9OXSZacbUzUZ6XIwXwYD
// SIG // VR0fBFgwVjBUoFKgUIZOaHR0cDovL3d3dy5taWNyb3Nv
// SIG // ZnQuY29tL3BraW9wcy9jcmwvTWljcm9zb2Z0JTIwVGlt
// SIG // ZS1TdGFtcCUyMFBDQSUyMDIwMTAoMSkuY3JsMGwGCCsG
// SIG // AQUFBwEBBGAwXjBcBggrBgEFBQcwAoZQaHR0cDovL3d3
// SIG // dy5taWNyb3NvZnQuY29tL3BraW9wcy9jZXJ0cy9NaWNy
// SIG // b3NvZnQlMjBUaW1lLVN0YW1wJTIwUENBJTIwMjAxMCgx
// SIG // KS5jcnQwDAYDVR0TAQH/BAIwADATBgNVHSUEDDAKBggr
// SIG // BgEFBQcDCDANBgkqhkiG9w0BAQsFAAOCAgEAuat+U9kS
// SIG // owBUTf93GI5zxc7T2uynqiSiZyoL4lI/SDHwvPxUNqMe
// SIG // lMMneNE0u8yr7bjJddRRZX1pBJ3Dd5aooYkgKiAzznVb
// SIG // 69aCWODRUATh8y6XrXqR5zINFC9lFx874yoNuFTiM29H
// SIG // bXT+vfkCMu0g6XFYfSiNBFyhtm6JaQcqakbbRkw3dCUJ
// SIG // DrTHYkqWg6Th8IAjDVxPMSLuPIYrk9O+aCKpxyYWt+/P
// SIG // w6boCNA2Y2HGR9XBvOCdOrnKucoK6HfbM69fTmwmSXSS
// SIG // 6cryVgsvD0QwJEYzyYJUpn59lEdCr56Mgjs47MwMDQqf
// SIG // ySZ5EadtMHiLc5ve4jYPhWRdSfRNenliAdbAq+1Y/UA+
// SIG // irfs/NaznC0X8YZTtv7CtPVMPg4kJD27ujmd07A7y7K/
// SIG // MhvVtgPMl3IIDaneF6S7XRJH+djiDdBoES0mNSzQweCC
// SIG // mPmY9+CMDGw27NrAooOS4tiC1hDFf+oMi9+ye9TW1u1c
// SIG // 6O33BqzZ3vTqsYBQtll+XXoMcid0sdmysOpW39OCml50
// SIG // wTks9EAvkXybs/JzqSH9vPlpjD4nF9pe4LPzKwuWp5si
// SIG // v523ipGjsfgF4cGC8xEzYPtSHMoI3cJ26limny3QjSn4
// SIG // bG/4DOUhW4A+QExF9WwMcFBVcSr0L9HSEQJuWg4xdw2a
// SIG // YL8t5kqI0qwTf/MwggdxMIIFWaADAgECAhMzAAAAFcXn
// SIG // a54Cm0mZAAAAAAAVMA0GCSqGSIb3DQEBCwUAMIGIMQsw
// SIG // CQYDVQQGEwJVUzETMBEGA1UECBMKV2FzaGluZ3RvbjEQ
// SIG // MA4GA1UEBxMHUmVkbW9uZDEeMBwGA1UEChMVTWljcm9z
// SIG // b2Z0IENvcnBvcmF0aW9uMTIwMAYDVQQDEylNaWNyb3Nv
// SIG // ZnQgUm9vdCBDZXJ0aWZpY2F0ZSBBdXRob3JpdHkgMjAx
// SIG // MDAeFw0yMTA5MzAxODIyMjVaFw0zMDA5MzAxODMyMjVa
// SIG // MHwxCzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5n
// SIG // dG9uMRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVN
// SIG // aWNyb3NvZnQgQ29ycG9yYXRpb24xJjAkBgNVBAMTHU1p
// SIG // Y3Jvc29mdCBUaW1lLVN0YW1wIFBDQSAyMDEwMIICIjAN
// SIG // BgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA5OGmTOe0
// SIG // ciELeaLL1yR5vQ7VgtP97pwHB9KpbE51yMo1V/YBf2xK
// SIG // 4OK9uT4XYDP/XE/HZveVU3Fa4n5KWv64NmeFRiMMtY0T
// SIG // z3cywBAY6GB9alKDRLemjkZrBxTzxXb1hlDcwUTIcVxR
// SIG // MTegCjhuje3XD9gmU3w5YQJ6xKr9cmmvHaus9ja+NSZk
// SIG // 2pg7uhp7M62AW36MEBydUv626GIl3GoPz130/o5Tz9bs
// SIG // hVZN7928jaTjkY+yOSxRnOlwaQ3KNi1wjjHINSi947SH
// SIG // JMPgyY9+tVSP3PoFVZhtaDuaRr3tpK56KTesy+uDRedG
// SIG // bsoy1cCGMFxPLOJiss254o2I5JasAUq7vnGpF1tnYN74
// SIG // kpEeHT39IM9zfUGaRnXNxF803RKJ1v2lIH1+/NmeRd+2
// SIG // ci/bfV+AutuqfjbsNkz2K26oElHovwUDo9Fzpk03dJQc
// SIG // NIIP8BDyt0cY7afomXw/TNuvXsLz1dhzPUNOwTM5TI4C
// SIG // vEJoLhDqhFFG4tG9ahhaYQFzymeiXtcodgLiMxhy16cg
// SIG // 8ML6EgrXY28MyTZki1ugpoMhXV8wdJGUlNi5UPkLiWHz
// SIG // NgY1GIRH29wb0f2y1BzFa/ZcUlFdEtsluq9QBXpsxREd
// SIG // cu+N+VLEhReTwDwV2xo3xwgVGD94q0W29R6HXtqPnhZy
// SIG // acaue7e3PmriLq0CAwEAAaOCAd0wggHZMBIGCSsGAQQB
// SIG // gjcVAQQFAgMBAAEwIwYJKwYBBAGCNxUCBBYEFCqnUv5k
// SIG // xJq+gpE8RjUpzxD/LwTuMB0GA1UdDgQWBBSfpxVdAF5i
// SIG // XYP05dJlpxtTNRnpcjBcBgNVHSAEVTBTMFEGDCsGAQQB
// SIG // gjdMg30BATBBMD8GCCsGAQUFBwIBFjNodHRwOi8vd3d3
// SIG // Lm1pY3Jvc29mdC5jb20vcGtpb3BzL0RvY3MvUmVwb3Np
// SIG // dG9yeS5odG0wEwYDVR0lBAwwCgYIKwYBBQUHAwgwGQYJ
// SIG // KwYBBAGCNxQCBAweCgBTAHUAYgBDAEEwCwYDVR0PBAQD
// SIG // AgGGMA8GA1UdEwEB/wQFMAMBAf8wHwYDVR0jBBgwFoAU
// SIG // 1fZWy4/oolxiaNE9lJBb186aGMQwVgYDVR0fBE8wTTBL
// SIG // oEmgR4ZFaHR0cDovL2NybC5taWNyb3NvZnQuY29tL3Br
// SIG // aS9jcmwvcHJvZHVjdHMvTWljUm9vQ2VyQXV0XzIwMTAt
// SIG // MDYtMjMuY3JsMFoGCCsGAQUFBwEBBE4wTDBKBggrBgEF
// SIG // BQcwAoY+aHR0cDovL3d3dy5taWNyb3NvZnQuY29tL3Br
// SIG // aS9jZXJ0cy9NaWNSb29DZXJBdXRfMjAxMC0wNi0yMy5j
// SIG // cnQwDQYJKoZIhvcNAQELBQADggIBAJ1VffwqreEsH2cB
// SIG // MSRb4Z5yS/ypb+pcFLY+TkdkeLEGk5c9MTO1OdfCcTY/
// SIG // 2mRsfNB1OW27DzHkwo/7bNGhlBgi7ulmZzpTTd2YurYe
// SIG // eNg2LpypglYAA7AFvonoaeC6Ce5732pvvinLbtg/SHUB
// SIG // 2RjebYIM9W0jVOR4U3UkV7ndn/OOPcbzaN9l9qRWqveV
// SIG // tihVJ9AkvUCgvxm2EhIRXT0n4ECWOKz3+SmJw7wXsFSF
// SIG // QrP8DJ6LGYnn8AtqgcKBGUIZUnWKNsIdw2FzLixre24/
// SIG // LAl4FOmRsqlb30mjdAy87JGA0j3mSj5mO0+7hvoyGtmW
// SIG // 9I/2kQH2zsZ0/fZMcm8Qq3UwxTSwethQ/gpY3UA8x1Rt
// SIG // nWN0SCyxTkctwRQEcb9k+SS+c23Kjgm9swFXSVRk2XPX
// SIG // fx5bRAGOWhmRaw2fpCjcZxkoJLo4S5pu+yFUa2pFEUep
// SIG // 8beuyOiJXk+d0tBMdrVXVAmxaQFEfnyhYWxz/gq77EFm
// SIG // PWn9y8FBSX5+k77L+DvktxW/tM4+pTFRhLy/AsGConsX
// SIG // HRWJjXD+57XQKBqJC4822rpM+Zv/Cuk0+CQ1ZyvgDbjm
// SIG // jJnW4SLq8CdCPSWU5nR0W2rRnj7tfqAxM328y+l7vzhw
// SIG // RNGQ8cirOoo6CGJ/2XBjU02N7oJtpQUQwXEGahC0HVUz
// SIG // WLOhcGbyoYIC1DCCAj0CAQEwggEAoYHYpIHVMIHSMQsw
// SIG // CQYDVQQGEwJVUzETMBEGA1UECBMKV2FzaGluZ3RvbjEQ
// SIG // MA4GA1UEBxMHUmVkbW9uZDEeMBwGA1UEChMVTWljcm9z
// SIG // b2Z0IENvcnBvcmF0aW9uMS0wKwYDVQQLEyRNaWNyb3Nv
// SIG // ZnQgSXJlbGFuZCBPcGVyYXRpb25zIExpbWl0ZWQxJjAk
// SIG // BgNVBAsTHVRoYWxlcyBUU1MgRVNOOjA4NDItNEJFNi1D
// SIG // MjlBMSUwIwYDVQQDExxNaWNyb3NvZnQgVGltZS1TdGFt
// SIG // cCBTZXJ2aWNloiMKAQEwBwYFKw4DAhoDFQB4d5NUBCgG
// SIG // RIbCxmBh7X/+gxuD86CBgzCBgKR+MHwxCzAJBgNVBAYT
// SIG // AlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQH
// SIG // EwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29y
// SIG // cG9yYXRpb24xJjAkBgNVBAMTHU1pY3Jvc29mdCBUaW1l
// SIG // LVN0YW1wIFBDQSAyMDEwMA0GCSqGSIb3DQEBBQUAAgUA
// SIG // 5htg5jAiGA8yMDIyMDUwMzE2MDU1OFoYDzIwMjIwNTA0
// SIG // MTYwNTU4WjB0MDoGCisGAQQBhFkKBAExLDAqMAoCBQDm
// SIG // G2DmAgEAMAcCAQACAiJeMAcCAQACAhF6MAoCBQDmHLJm
// SIG // AgEAMDYGCisGAQQBhFkKBAIxKDAmMAwGCisGAQQBhFkK
// SIG // AwKgCjAIAgEAAgMHoSChCjAIAgEAAgMBhqAwDQYJKoZI
// SIG // hvcNAQEFBQADgYEAIIRGBvXkZKIp85cHnZ+fzS+8qqVe
// SIG // RY42sWvPpCppdqZr93eIV1idwK894nJlFnjqxesozJ4h
// SIG // bJzGnbTsuufGtvu0b83qR0WcjDaZP5NMx6uyGN3pxLCQ
// SIG // JsIWVkeIsWiXtdPU39XWy2naB9NlPZ1zre2H5WHAXqNe
// SIG // IfeCLdNEv6QxggQNMIIECQIBATCBkzB8MQswCQYDVQQG
// SIG // EwJVUzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UE
// SIG // BxMHUmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENv
// SIG // cnBvcmF0aW9uMSYwJAYDVQQDEx1NaWNyb3NvZnQgVGlt
// SIG // ZS1TdGFtcCBQQ0EgMjAxMAITMwAAAYdCFmYEXPP0jQAB
// SIG // AAABhzANBglghkgBZQMEAgEFAKCCAUowGgYJKoZIhvcN
// SIG // AQkDMQ0GCyqGSIb3DQEJEAEEMC8GCSqGSIb3DQEJBDEi
// SIG // BCDXPm8HZZ5nTVivejGizToVXtgXOc5qJgsWVX4riGoI
// SIG // 3zCB+gYLKoZIhvcNAQkQAi8xgeowgecwgeQwgb0EIMQs
// SIG // 8KATbqAfdsLAtksWYc+B1RsrwmZ8GDbnM6RsmptYMIGY
// SIG // MIGApH4wfDELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldh
// SIG // c2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNV
// SIG // BAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEmMCQGA1UE
// SIG // AxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAgUENBIDIwMTAC
// SIG // EzMAAAGHQhZmBFzz9I0AAQAAAYcwIgQgAzwEqLncWA5a
// SIG // PbnQi8B1DxE0V1Ox5dnK++U2LZCfFCAwDQYJKoZIhvcN
// SIG // AQELBQAEggIAMDzLylk6xil3PFF9bYpBq0hoyobfQpsM
// SIG // gXBXpQIzRPcU2y8jMx35kORoPY1bzAE8Ntg4cMrjdRon
// SIG // Plri0k2Dl+5QQtGwKy1z+Yc4s60HcNOF3/LMQtWJFw9E
// SIG // K2HdFuuno5X8avzRXFVoJXjn/84XtJI/zuBIpJYPhMhY
// SIG // AXX009b66vF9Jeli2Tmc/KR2hSiaAQTATV6LrtEF6qw+
// SIG // ooqaZEa4Kxw8OJYT95zVQLXBwJS88JqOf622pE1O9mWQ
// SIG // P+4AyJRcobH9q1AxwVo/9K004u+d7pH5FGS9CxZSaWIK
// SIG // N9wUjue58duXyuMBiyRdAEUNyozOb/DhRhXYWPUQNcg7
// SIG // CT9I8U5TFT1h1ju51Ti8KLyk+tsr3WrpfqAXbRhb04Wq
// SIG // 2936qP6QqMKPj0xYxfyY+/ByAsAWAOO315MjEiGxkTai
// SIG // gcZT5/sP+lcLi1KTckTGILoFYMzfJDPsNFCABsiKlgh8
// SIG // QUGRQ68dIEnwAH9eqYMPAS9lDVR5j4Zf9ghjTqpDg+qV
// SIG // RFEFgVau0Tibjscj+5ts9OxnbmnpCdNTb07u7E2fFQsW
// SIG // 1jugGLRfy2kHViBtKfKbb3VasDgQGQ38964QCmi7oSVC
// SIG // JMCvGOfPCfEWDHi4G/xWY7oUWUeeksF6Xn2RI9f6s3ZI
// SIG // k+Llw7vNbo8aWTec9Oe8GO54qKQl7KbgjM4=
// SIG // End signature block
